File: DecryptBackup.java
import java.io.*;
import java.nio.file.*;
import java.security.spec.KeySpec;
import java.util.Base64;
import javax.crypto.Cipher;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.PBEKeySpec;
import javax.crypto.spec.SecretKeySpec;
import java.time.LocalDateTime;

/**
 * SecureVault Backup Decryption Utility
 * Decrypts SecureVault backup files and extracts password data
 */
public class DecryptBackup {

    // Encryption constants (must match BackupEncryption.kt)
    private static final String ALGORITHM = "AES";
    private static final String TRANSFORMATION = "AES/CBC/PKCS5Padding";
    private static final String KEY_DERIVATION_ALGORITHM = "PBKDF2WithHmacSHA256";
    private static final int KEY_LENGTH = 256;
    private static final int IV_LENGTH = 16;
    private static final int SALT_LENGTH = 32;
    private static final int ITERATION_COUNT = 100000;

    private static SecretKeySpec deriveKey(String password, byte[] salt) throws Exception {
        SecretKeyFactory factory = SecretKeyFactory.getInstance(KEY_DERIVATION_ALGORITHM);
        KeySpec spec = new PBEKeySpec(password.toCharArray(), salt, ITERATION_COUNT, KEY_LENGTH);
        byte[] key = factory.generateSecret(spec).getEncoded();
        return new SecretKeySpec(key, ALGORITHM);
    }

    private static String decrypt(String encryptedData, String password) throws Exception {
        // Decode from Base64
        byte[] combined = Base64.getDecoder().decode(encryptedData);

        if (combined.length < SALT_LENGTH + IV_LENGTH) {
            throw new Exception("Invalid encrypted data format");
        }

        // Extract salt, IV, and encrypted data
        byte[] salt = new byte[SALT_LENGTH];
        byte[] iv = new byte[IV_LENGTH];
        byte[] encrypted = new byte[combined.length - SALT_LENGTH - IV_LENGTH];

        System.arraycopy(combined, 0, salt, 0, SALT_LENGTH);
        System.arraycopy(combined, SALT_LENGTH, iv, 0, IV_LENGTH);
        System.arraycopy(combined, SALT_LENGTH + IV_LENGTH, encrypted, 0, encrypted.length);

        // Derive key from password using same salt
        SecretKeySpec key = deriveKey(password, salt);

        // Decrypt the data
        Cipher cipher = Cipher.getInstance(TRANSFORMATION);
        cipher.init(Cipher.DECRYPT_MODE, key, new IvParameterSpec(iv));
        byte[] decryptedData = cipher.doFinal(encrypted);

        return new String(decryptedData, "UTF-8");
    }

    public static void main(String[] args) {
        String backupFile = args.length > 0 ? args[0] :
            "/Users/akshitharsola/Documents/SecureAttend/SecureVault-Android-master/app/SecureVault_Backup_2026-01-05_18-41-22.backup";
        String password = args.length > 1 ? args[1] : "147zaq";
        String outputFile = args.length > 2 ? args[2] : "decrypted_passwords.txt";

        System.out.println("=".repeat(70));
        System.out.println("SecureVault Backup Decryption Script");
        System.out.println("=".repeat(70));
        System.out.println("Backup file: " + backupFile);
        System.out.println("Output file: " + outputFile);
        System.out.println();

        try {
            // Read backup file
            System.out.println("[1/4] Reading backup file...");
            String backupContent = Files.readString(Paths.get(backupFile));
            System.out.println("✓ Backup file read successfully (" + backupContent.length() + " bytes)");

            // Parse JSON to extract encrypted data
            System.out.println("\n[2/4] Parsing backup JSON structure...");

            String encryptedData;
            if (backupContent.trim().startsWith("{")) {
                // JSON format - extract "data" or "f" field using proper JSON parsing
                // Look for the pattern "data": "..." or "f": "..."
                String pattern = backupContent.contains("\"data\"") ? "\"data\"" : "\"f\"";

                int fieldStart = backupContent.indexOf(pattern);
                if (fieldStart == -1) {
                    throw new Exception("No 'data' or 'f' field found in backup JSON");
                }

                // Find the start of the value (after ": ")
                int valueStart = backupContent.indexOf("\"", fieldStart + pattern.length()) + 1;

                // Find the end of the value, handling escaped quotes and unicode escapes
                StringBuilder sb = new StringBuilder();
                boolean inEscape = false;
                for (int i = valueStart; i < backupContent.length(); i++) {
                    char c = backupContent.charAt(i);

                    if (inEscape) {
                        // Handle escape sequences
                        if (c == 'u') {
                            // Unicode escape like \u003d
                            String unicode = backupContent.substring(i + 1, i + 5);
                            sb.append((char) Integer.parseInt(unicode, 16));
                            i += 4; // Skip the next 4 characters
                        } else if (c == 'n') {
                            sb.append('\n');
                        } else if (c == 'r') {
                            sb.append('\r');
                        } else if (c == 't') {
                            sb.append('\t');
                        } else if (c == '\\') {
                            sb.append('\\');
                        } else if (c == '"') {
                            sb.append('"');
                        } else {
                            sb.append(c);
                        }
                        inEscape = false;
                    } else if (c == '\\') {
                        inEscape = true;
                    } else if (c == '"') {
                        // End of string value
                        break;
                    } else {
                        sb.append(c);
                    }
                }

                encryptedData = sb.toString();
                String formatType = pattern.equals("\"data\"") ? "new format" : "legacy obfuscated format";
                System.out.println("✓ Found " + pattern + " field (" + formatType + ") - " + encryptedData.length() + " bytes");
            } else {
                // Very old format - entire file is encrypted data
                System.out.println("✓ Detected very old backup format (raw encrypted data)");
                encryptedData = backupContent.trim();
            }

            // Decrypt the data
            System.out.println("\n[3/4] Decrypting password data...");
            String decryptedJson = decrypt(encryptedData, password);
            System.out.println("✓ Decryption successful!");

            // Count passwords
            int passwordCount = 0;
            int idx = 0;
            while ((idx = decryptedJson.indexOf("{", idx + 1)) != -1) {
                passwordCount++;
            }
            System.out.println("✓ Found approximately " + passwordCount + " password entries");

            // Write to output file
            System.out.println("\n[4/4] Writing decrypted data to " + outputFile + "...");
            StringBuilder output = new StringBuilder();
            output.append("=".repeat(70)).append("\n");
            output.append("SecureVault Decrypted Backup Data\n");
            output.append("=".repeat(70)).append("\n");
            output.append("Backup file: ").append(backupFile).append("\n");
            output.append("Decrypted at: ").append(LocalDateTime.now()).append("\n");
            output.append("Password count: ~").append(passwordCount).append("\n");
            output.append("=".repeat(70)).append("\n\n");
            output.append("RAW DECRYPTED JSON:\n");
            output.append("-".repeat(70)).append("\n");
            output.append(decryptedJson).append("\n");
            output.append("-".repeat(70)).append("\n");

            Files.writeString(Paths.get(outputFile), output.toString());
            System.out.println("✓ Data written successfully!");

            System.out.println("\n" + "=".repeat(70));
            System.out.println("SUCCESS! Decrypted data saved to: " + outputFile);
            System.out.println("=".repeat(70));

            // Print preview
            System.out.println("\nPreview of decrypted JSON:");
            System.out.println("-".repeat(70));
            String preview = decryptedJson.length() > 1000 ?
                decryptedJson.substring(0, 1000) + "\n..." :
                decryptedJson;
            System.out.println(preview);
            System.out.println("-".repeat(70));

        } catch (Exception e) {
            System.out.println("\n" + "=".repeat(70));
            System.out.println("ERROR: " + e.getMessage());
            System.out.println("=".repeat(70));
            e.printStackTrace();
            System.exit(1);
        }
    }
}

File: tools/decrypt_backup.py
#!/usr/bin/env python3
"""
SecureVault Backup Decryption Script
Decrypts SecureVault backup files and extracts password data
"""

import sys
import json
import base64
import re
from datetime import datetime
from Crypto.Cipher import AES
from Crypto.Protocol.KDF import PBKDF2
from Crypto.Hash import SHA256

# Encryption constants (must match BackupEncryption.kt)
KEY_LENGTH = 32  # 256 bits / 8
IV_LENGTH = 16
SALT_LENGTH = 32
ITERATION_COUNT = 100000

def derive_key(password, salt):
    """Derives encryption key from password using PBKDF2"""
    return PBKDF2(
        password.encode('utf-8'),
        salt,
        dkLen=KEY_LENGTH,
        count=ITERATION_COUNT,
        hmac_hash_module=SHA256
    )

def decrypt(encrypted_data_b64, password):
    """Decrypts data using password-based encryption"""
    # Decode from Base64
    combined = base64.b64decode(encrypted_data_b64)

    if len(combined) < SALT_LENGTH + IV_LENGTH:
        raise ValueError("Invalid encrypted data format")

    # Extract salt, IV, and encrypted data
    salt = combined[0:SALT_LENGTH]
    iv = combined[SALT_LENGTH:SALT_LENGTH + IV_LENGTH]
    encrypted = combined[SALT_LENGTH + IV_LENGTH:]

    # Derive key from password using same salt
    key = derive_key(password, salt)

    # Decrypt the data
    cipher = AES.new(key, AES.MODE_CBC, iv)
    decrypted_data = cipher.decrypt(encrypted)

    # Remove PKCS5 padding
    padding_length = decrypted_data[-1]
    decrypted_data = decrypted_data[:-padding_length]

    return decrypted_data.decode('utf-8')

def main():
    backup_file = sys.argv[1] if len(sys.argv) > 1 else "/Users/akshitharsola/Documents/SecureAttend/SecureVault-Android-master/app/SecureVault_Backup_2026-01-05_18-41-22.backup"
    password = sys.argv[2] if len(sys.argv) > 2 else "147zaq"
    output_file = sys.argv[3] if len(sys.argv) > 3 else "decrypted_passwords.txt"

    print("=" * 70)
    print("SecureVault Backup Decryption Script")
    print("=" * 70)
    print(f"Backup file: {backup_file}")
    print(f"Output file: {output_file}")
    print()

    try:
        # Read backup file
        print("[1/4] Reading backup file...")
        with open(backup_file, 'r') as f:
            backup_content = f.read()
        print(f"✓ Backup file read successfully ({len(backup_content)} bytes)")

        # Parse JSON to extract encrypted data
        print("\n[2/4] Parsing backup JSON structure...")

        # Try to parse as JSON first
        try:
            backup_json = json.loads(backup_content)

            # Check for different format keys
            if 'data' in backup_json:
                encrypted_data = backup_json['data']
                print(f"✓ Found 'data' field (new format) - {len(encrypted_data)} bytes")
            elif 'f' in backup_json:
                encrypted_data = backup_json['f']
                print(f"✓ Found 'f' field (legacy obfuscated format) - {len(encrypted_data)} bytes")
            else:
                raise ValueError("No 'data' or 'f' field found in backup JSON")

        except json.JSONDecodeError:
            # Very old format - entire file is encrypted data
            print("✓ Detected very old backup format (raw encrypted data)")
            encrypted_data = backup_content.strip()

        # Decrypt the data
        print("\n[3/4] Decrypting password data...")
        decrypted_json = decrypt(encrypted_data, password)
        print("✓ Decryption successful!")

        # Parse decrypted JSON to pretty print it
        try:
            passwords = json.loads(decrypted_json)
            pretty_json = json.dumps(passwords, indent=2)
            password_count = len(passwords) if isinstance(passwords, list) else "unknown"
            print(f"✓ Found {password_count} password entries")
        except:
            pretty_json = decrypted_json
            password_count = "unknown"

        # Write to output file
        print(f"\n[4/4] Writing decrypted data to {output_file}...")
        with open(output_file, 'w') as f:
            f.write("=" * 70 + "\n")
            f.write("SecureVault Decrypted Backup Data\n")
            f.write("=" * 70 + "\n")
            f.write(f"Backup file: {backup_file}\n")
            f.write(f"Decrypted at: {datetime.now()}\n")
            f.write(f"Password count: {password_count}\n")
            f.write("=" * 70 + "\n\n")
            f.write("RAW DECRYPTED JSON:\n")
            f.write("-" * 70 + "\n")
            f.write(pretty_json)
            f.write("\n" + "-" * 70 + "\n")

        print("✓ Data written successfully!")

        print("\n" + "=" * 70)
        print(f"SUCCESS! Decrypted data saved to: {output_file}")
        print("=" * 70)

        # Print preview
        print("\nPreview of decrypted JSON:")
        print("-" * 70)
        preview = pretty_json[:1000]
        print(preview + ("..." if len(pretty_json) > 1000 else ""))
        print("-" * 70)

    except Exception as e:
        print("\n" + "=" * 70)
        print(f"ERROR: {str(e)}")
        print("=" * 70)
        import traceback
        traceback.print_exc()
        sys.exit(1)

if __name__ == "__main__":
    main()
